src/listen.js

```diff
@@ -0,0 +1,49 @@
+'use strict'
+const PeerInfo = require('peer-info')
+const PeerId = require('peer-id')
+const multiaddr = require('multiaddr')
+const pull = require('pull-stream')
+const lp = require('pull-length-prefixed')
+
+const msg = require('./message')
+
+module.exports = (conn, callback) => {
+  pull(
+    conn,
+    lp.decode(),
+    pull.take(1),
+    pull.collect((err, data) => {
+      if (err) {
+        return callback(err)
+      }
+
+      const input = msg.decode(data[0])
+
+      const id = PeerId.createFromPubKey(input.publicKey)
+      const info = new PeerInfo(id)
+      input.listenAddrs
+        .map(multiaddr)
+        .forEach((ma) => info.multiaddr.add(ma))
+
+      callback(null, info, getObservedAddrs(input))
+    })
+  )
```

Comment:
You can imagine `collect`, like using `bl`, so it is actually called when the `conn` is ended.